ratelimit/src/lib.rs

```diff
@@ -1,203 +1,470 @@
+            }
+            (None, Some(interval)) => {
+                let a = self.refill_rate.0;
+                let i = self.refill_rate.1.as_nanos() as u64;
+
+                // tokens per nanosecond
+                let target_rate = a as f64 / i as f64;
+
+                let amount = (interval.as_nanos() as u64 as f64 * target_rate).round() as u64;
+                let interval = interval.as_nanos() as u64;
+
+                (amount, interval)
+            }
+            (None, None) => {
+                // work out an amount and interval based on the rate
+
+                let mut amount = self.refill_rate.0;
+                let mut interval = self.refill_rate.1.as_nanos() as u64;
+
+                // copy the original values for use later
+                let a = amount;
+                let i = interval;
+
+                // reduce things to the smallest amount and interval possible
+                let gcd = gcd(amount, interval);
+                amount /= gcd;
+                interval /= gcd;
+
+                // scale things up to reduce bookeeping
+                while interval < 100 {
+                    amount += a;
+                    interval += i;
                 }
+
+                (amount, interval)
             }
+        };
+
+        let available = AtomicU64::new(self.available);
+        let capacity = AtomicU64::new(std::cmp::max(self.capacity, amount));
+
+        let refill_amount = AtomicU64::new(amount);
+        let refill_at = AtomicInstant::new(
+            Instant::now() + clocksource::Duration::<Nanoseconds<u64>>::from_nanos(interval),
+        );
+        let refill_interval = AtomicDuration::from_nanos(interval);
+
+        Ratelimiter {
+            available,
+            capacity,
+            refill_amount,
+            refill_at,
+            refill_interval,
         }
     }
+}
 
-    /// Non-blocking wait for one token, returns an `Ok` if a token was
-    /// successfully acquired, returns an `Err` if it would block.
-    ///
-    /// # Examples
-    ///
-    /// ```
-    /// use ratelimit::*;
-    ///
-    /// let ratelimiter = Ratelimiter::new(1, 1, 100);
-    /// for i in 0..100 {
-    ///     // do some work here
-    ///     while ratelimiter.try_wait().is_err() {
-    ///         // do some other work
-    ///     }
-    /// }
-    /// ```
-    pub fn try_wait(&self) -> Result<(), std::io::Error> {
-        self.tick();
-        let mut current = self.available.load(Ordering::Relaxed);
-        while current > 0 {
-            match self.available.compare_exchange(
-                current,
-                current - 1,
-                Ordering::Relaxed,
-                Ordering::Relaxed,
-            ) {
-                Ok(_) => return Ok(()),
-                Err(v) => {
-                    current = v;
-                }
+// Taken from Wikipedia: https://en.wikipedia.org/wiki/Binary_GCD_algorithm
+fn gcd(mut u: u64, mut v: u64) -> u64 {
+    use std::cmp::min;
```

Comment:
What about the `Gcd` crate?

Reply:
same. require explicit amount / interval values now.